src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -727,6 +728,89 @@ bool is_object_additional_properties_superset(
       get_additional_props(newer));
 }
 
+bool is_object_properties_superset(
+  json::Value const& older, json::Value const& newer) {
+    // check that every property in newer["properties"]
+    // if it appears in older["properties"],
+    //    then it has to be compatible with the schema
+    // or if for every match with a pattern in older["patternProperties"],
+    //    then it has to be compatible with the schema,
+    // or
+    //    it has to be compatible with older["additionalProperties"]
+
+    auto newer_properties = get_object_or_empty(newer, "properties");
+    if (newer_properties.ObjectEmpty()) {
+        // no "properties" in newer, all good
+        return true;
+    }
+
+    // older["properties"] is a map of <prop, schema>
+    auto older_properties = get_object_or_empty(older, "properties");
+    // older["patternProperties"] is a map of <pattern, schema>
+    auto older_pattern_properties = get_object_or_empty(
+      older, "patternProperties");
+    // older["additionalProperties"] is a schema
+    auto& older_additional_properties = [&]() -> json::Value const& {
+        auto older_it = older.FindMember("additionalProperties");
+        if (older_it == older.MemberEnd()) {
+            // default is `true`
+            return get_true_schema();
+        }
+
+        if (older_it->value.IsBool()) {
+            return older_it->value.GetBool() ? get_true_schema()
+                                             : get_false_schema();
+        }
+
+        return older_it->value;
+    }();
```

Comment:
nitpick: The invocation can be deferred so that it's not called if it's not needed.

Reply:
good idea, done